File: packages/sitecompiler-py/sitecompiler/resources/jobs.py
import time
import httpx
from typing import Optional, Callable
from ..models import Job, ExportFormat
from ..exceptions import SiteCompilerError, JobNotFoundError, TimeoutError

class JobsResource:
    def __init__(self, client: httpx.Client, base_url: str):
        self._client = client
        self._base_url = base_url

    def create(self, url: str, format: ExportFormat = "nextjs", idempotency_key: Optional[str] = None) -> Job:
        headers = {}
        if idempotency_key:
            headers["x-idempotency-key"] = idempotency_key

        res = self._client.post(
            f"{self._base_url}/api/export",
            json={"url": url, "format": format},
            headers=headers,
        )
        if not res.is_success:
            raise SiteCompilerError(f"Failed to create job: {res.text}", res.status_code)

        data = res.json()
        job_id = data.get("jobId") or data.get("id")
        return self.get(job_id)

    def get(self, job_id: str) -> Job:
        res = self._client.get(f"{self._base_url}/api/job/{job_id}/status")
        if res.status_code == 404:
            raise JobNotFoundError(f"Job {job_id} not found", 404)
        if not res.is_success:
            raise SiteCompilerError(f"Failed to fetch job status: {res.text}", res.status_code)
        return Job(**res.json())

    def cancel(self, job_id: str) -> Job:
        res = self._client.post(f"{self._base_url}/api/job/{job_id}/cancel")
        if not res.is_success:
            raise SiteCompilerError(f"Failed to cancel job: {res.text}", res.status_code)
        return Job(**res.json())

    def poll_until_complete(
        self,
        job_id: str,
        interval_seconds: float = 2.0,
        timeout_seconds: float = 900.0,
        on_progress: Optional[Callable[[Job], None]] = None,
    ) -> Job:
        start_time = time.time()
        while time.time() - start_time < timeout_seconds:
            job = self.get(job_id)
            if on_progress:
                on_progress(job)

            if job.status in ["completed", "failed", "cancelled"]:
                return job

            time.sleep(interval_seconds)

        raise TimeoutError(f"Job {job_id} did not complete within {timeout_seconds} seconds")

File: packages/sitecompiler-sdk/src/resources/jobs.ts
import { CreateJobParams, Job, PollOptions } from '../types';
import { JobNotFoundError, SiteCompilerError, TimeoutError } from '../errors';

export class JobsResource {
  constructor(
    private readonly baseUrl: string,
    private readonly fetchFn: typeof fetch,
    private readonly headers: Record<string, string>
  ) {}

  /**
   * Submit a new website compilation job
   */
  async create(params: CreateJobParams): Promise<Job> {
    const res = await this.fetchFn(`${this.baseUrl}/api/export`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.headers,
        ...(params.idempotencyKey ? { 'x-idempotency-key': params.idempotencyKey } : {}),
      },
      body: JSON.stringify({
        url: params.url,
        format: params.format || 'nextjs',
      }),
    });

    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new SiteCompilerError(errorData.error || 'Failed to create export job', res.status, errorData);
    }

    const data = await res.json();
    return this.get(data.jobId || data.id);
  }

  /**
   * Retrieve the status and logs of an ongoing or completed job
   */
  async get(jobId: string): Promise<Job> {
    const res = await this.fetchFn(`${this.baseUrl}/api/job/${jobId}/status`, {
      headers: this.headers,
    });

    if (res.status === 404) {
      throw new JobNotFoundError(jobId);
    }

    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new SiteCompilerError(errorData.error || 'Failed to fetch job status', res.status, errorData);
    }

    return res.json();
  }

  /**
   * Cancel an active job
   */
  async cancel(jobId: string): Promise<Job> {
    const res = await this.fetchFn(`${this.baseUrl}/api/job/${jobId}/cancel`, {
      method: 'POST',
      headers: this.headers,
    });

    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new SiteCompilerError(errorData.error || 'Failed to cancel job', res.status, errorData);
    }

    return res.json();
  }

  /**
   * Poll a job until it reaches a terminal status (completed, failed, cancelled)
   */
  async pollUntilComplete(jobId: string, options: PollOptions = {}): Promise<Job> {
    const intervalMs = options.intervalMs || 2000;
    const timeoutMs = options.timeoutMs || 900000; // 15 mins default
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      const job = await this.get(jobId);
      if (options.onProgress) {
        options.onProgress(job);
      }

      if (job.status === 'completed' || job.status === 'failed' || job.status === 'cancelled') {
        return job;
      }

      await new Promise((r) => setTimeout(r, intervalMs));
    }

    throw new TimeoutError(`Job ${jobId} did not complete within ${timeoutMs}ms`);
  }
}
